server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +268,198 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
     return paContinue;
 }
 
-void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut) {
+std::string SC_PortAudioDriver::GetPaDeviceName(int index) {
     const PaDeviceInfo* pdi;
+    pdi = Pa_GetDeviceInfo(index);
+    std::string name;
+#ifndef __APPLE__
     const PaHostApiInfo* apiInfo;
-    char devString[256];
+    apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
+    name.append(apiInfo->name);
+    name.append(" : ");
+#endif
+    name.append(pdi->name);
+    return name;
+}
+
+void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut, IOType ioType) {
+    const PaDeviceInfo* pdi;
     PaDeviceIndex numDevices = Pa_GetDeviceCount();
-    mInOut[0] = paNoDevice;
-    mInOut[1] = paNoDevice;
+    *mInOut = paNoDevice;
+
+    if (device) {
+        for (int i = 0; i < numDevices; i++) {
+            pdi = Pa_GetDeviceInfo(i);
+            std::string devString = GetPaDeviceName(i);
+            // compare strings, but only if the string is not empty
+            if (strstr(devString.c_str(), device) && device[0]) {
+                if (ioType == IOType::Input && pdi->maxInputChannels > 0) {
+                    *mInOut = i;
+                    break;
+                } else if (ioType == IOType::Output && pdi->maxOutputChannels > 0) {
+                    *mInOut = i;
+                    break;
+                }
+            }
+        }
+    }
+}
 
-    // This tries to find one or two devices that match the given name (substring)
-    // might cause problems for some names...
-    for (int i = 0; i < numDevices; i++) {
-        pdi = Pa_GetDeviceInfo(i);
-        apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
-        strcpy(devString, apiInfo->name);
-        strcat(devString, " : ");
-        strcat(devString, pdi->name);
-        if (strstr(devString, device)) {
-            if (pdi->maxInputChannels > 0)
-                mInOut[0] = i;
-            if (pdi->maxOutputChannels > 0)
-                mInOut[1] = i;
+PaStreamParameters SC_PortAudioDriver::GetPaStreamParameters(int* device, int channelCount, double suggestedLatency) {
+    PaStreamParameters streamParams;
+    PaSampleFormat fmt = paFloat32 | paNonInterleaved;
+    streamParams.device = *device;
+    streamParams.channelCount = channelCount;
+    streamParams.sampleFormat = fmt;
+    if (suggestedLatency)
+        streamParams.suggestedLatency = suggestedLatency;
```

Comment:
just assign this directly -- this is a POD (plain old data, i.e. no constructor) so if you don't assign anything to suggestedLatency it will be uninitialized and garbage

Reply:
additionally, the first function parameter can be plain int, no need to pass a pointer